File: app/src/main/java/com/example/myapplication/core/profile/ProfileModels.kt
package com.example.myapplication.core.profile

import java.time.DateTimeException
import java.time.LocalDate
import java.time.Period

enum class MetabolicSex {
    FEMALE,
    MALE,
}

enum class ActivityLevel(val multiplier: Double) {
    SEDENTARY(1.20),
    LIGHT(1.375),
    MODERATE(1.55),
    HIGH(1.725),
}

enum class GoalPace {
    MILD,
    STANDARD,
    AGGRESSIVE,
}

data class PersonalProfile(
    val birthDateEpochDay: Long,
    val metabolicSex: MetabolicSex,
    val heightCm: Double,
    val currentWeightKg: Double,
    val targetWeightKg: Double,
    val activityLevel: ActivityLevel,
    val goalPace: GoalPace,
    val personalizationConsent: Boolean,
    val cloudAiConsent: Boolean,
) {
    fun validationIssues(today: LocalDate): List<String> = buildList {
        val birthDate = try {
            LocalDate.ofEpochDay(birthDateEpochDay)
        } catch (_: DateTimeException) {
            null
        }
        val age = birthDate?.let { Period.between(it, today).years }

        if (age == null || age !in MINIMUM_AGE..MAXIMUM_AGE) {
            add("Độ tuổi phải từ 18 đến 100.")
        }
        if (!heightCm.isFinite() || heightCm !in MINIMUM_HEIGHT_CM..MAXIMUM_HEIGHT_CM) {
            add("Chiều cao phải từ 100 đến 250 cm.")
        }
        if (!currentWeightKg.isFinite() || currentWeightKg !in MINIMUM_WEIGHT_KG..MAXIMUM_WEIGHT_KG) {
            add("Cân nặng hiện tại phải từ 30 đến 350 kg.")
        }
        if (!targetWeightKg.isFinite() || targetWeightKg !in MINIMUM_WEIGHT_KG..MAXIMUM_WEIGHT_KG) {
            add("Cân nặng mục tiêu phải từ 30 đến 350 kg.")
        }
        if (!personalizationConsent) {
            add("Bạn cần đồng ý sử dụng dữ liệu hồ sơ để bật cá nhân hóa.")
        }
    }

    private companion object {
        const val MINIMUM_AGE = 18
        const val MAXIMUM_AGE = 100
        const val MINIMUM_HEIGHT_CM = 100.0
        const val MAXIMUM_HEIGHT_CM = 250.0
        const val MINIMUM_WEIGHT_KG = 30.0
        const val MAXIMUM_WEIGHT_KG = 350.0
    }
}

File: flutter/lib/core/model/profile_models.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'profile_models.freezed.dart';
part 'profile_models.g.dart';

enum MetabolicSex {
  @JsonValue('FEMALE')
  female,
  @JsonValue('MALE')
  male,
}

enum ActivityLevel {
  @JsonValue('SEDENTARY')
  sedentary(1.20),
  @JsonValue('LIGHT')
  light(1.375),
  @JsonValue('MODERATE')
  moderate(1.55),
  @JsonValue('HIGH')
  high(1.725);

  final double multiplier;
  const ActivityLevel(this.multiplier);
}

enum GoalPace {
  @JsonValue('MILD')
  mild,
  @JsonValue('STANDARD')
  standard,
  @JsonValue('AGGRESSIVE')
  aggressive,
}

@freezed
abstract class PersonalProfile with _$PersonalProfile {
  const PersonalProfile._();

  const factory PersonalProfile({
    required int birthDateEpochDay,
    required MetabolicSex metabolicSex,
    required double heightCm,
    required double currentWeightKg,
    required double targetWeightKg,
    required ActivityLevel activityLevel,
    required GoalPace goalPace,
    required bool personalizationConsent,
    required bool cloudAiConsent,
  }) = _PersonalProfile;

  factory PersonalProfile.fromJson(Map<String, dynamic> json) =>
      _$PersonalProfileFromJson(json);

  List<String> validationIssues(DateTime today) {
    final issues = <String>[];
    
    // Convert epoch day to DateTime
    final birthDate = DateTime.fromMillisecondsSinceEpoch(
      birthDateEpochDay * 24 * 60 * 60 * 1000,
      isUtc: true,
    );
    
    // Calculate age
    int age = today.year - birthDate.year;
    if (today.month < birthDate.month ||
        (today.month == birthDate.month && today.day < birthDate.day)) {
      age--;
    }

    if (age < _minimumAge || age > _maximumAge) {
      issues.add("Độ tuổi phải từ 18 đến 100.");
    }
    if (!heightCm.isFinite ||
        heightCm < _minimumHeightCm ||
        heightCm > _maximumHeightCm) {
      issues.add("Chiều cao phải từ 100 đến 250 cm.");
    }
    if (!currentWeightKg.isFinite ||
        currentWeightKg < _minimumWeightKg ||
        currentWeightKg > _maximumWeightKg) {
      issues.add("Cân nặng hiện tại phải từ 30 đến 350 kg.");
    }
    if (!targetWeightKg.isFinite ||
        targetWeightKg < _minimumWeightKg ||
        targetWeightKg > _maximumWeightKg) {
      issues.add("Cân nặng mục tiêu phải từ 30 đến 350 kg.");
    }
    if (!personalizationConsent) {
      issues.add("Bạn cần đồng ý sử dụng dữ liệu hồ sơ để bật cá nhân hóa.");
    }

    return issues;
  }

  static const int _minimumAge = 18;
  static const int _maximumAge = 100;
  static const double _minimumHeightCm = 100.0;
  static const double _maximumHeightCm = 250.0;
  static const double _minimumWeightKg = 30.0;
  static const double _maximumWeightKg = 350.0;
}
